sdl_android/src/main/java/com/smartdevicelink/proxy/SdlProxyBase.java

```diff
@@ -3768,27 +3808,113 @@ public Surface createOpenGLInputSurface(int frameRate, int iFrameInterval, int w
+            scheduler.execute(fTask);
+
+            while (!navServiceStartResponseReceived && !fTask.isDone());
+            scheduler.shutdown();
+            scheduler = null;
+            fTask = null;
+
+            if (navServiceStartResponse) {
+                return sdlSession.createOpenGLInputSurface(frameRate, iFrameInterval, width,
+                        height, bitrate, SessionType.NAV, sdlSession.getSessionId());
+            }
+
+            if (navServiceStartRejectedParams != null) {
+                StringBuilder builder = new StringBuilder();
+                for (String paramName : navServiceStartRejectedParams) {
+                    if (builder.length() > 0) {
+                        builder.append(", ");
+                    }
+                    builder.append(paramName);
+                }
+                DebugTool.logWarning("StartService for nav failed. Rejected params: " + builder.toString());
+
+                if (!navServiceStartRejectedParams.contains(ControlFrameTags.Video.StartService.VIDEO_PROTOCOL)
+                        && !navServiceStartRejectedParams.contains(ControlFrameTags.Video.StartService.VIDEO_CODEC)) {
+                    // The reason of NACK is not protocol nor codec. There is no point retrying with
+                    // another video format, so we simply fail here.
+                    break;
+                }
+            } else {
+                DebugTool.logWarning("StartService for nav failed (rejected params not supplied)");
+            }
         }
+        return null;
     }
-    
+
+    /**
+     * Creates a list of VideoStreamingParams with video formats that are supported by both HMI and proxy
+     *
+     * Note: this method does not take care of matching video resolution. App should look into
+     * HMI capability's preferredResolution and adjust width and height accordingly.
+     *
+     * @param hmiCapability HMI's capability information
+     * @param frameRate specified rate of frames
+     * @param frameInterval specified interval
+     * @param width specified width of the video
+     * @param height specified height of the video
+     * @param bitrate specified bitrate of the video
+     * @param availableFormats list of video formats supported by proxy
+     * @return list of VideoStreamingParams instance. This list can be empty.
+     */
+    private static List<VideoStreamingParams> createDesiredVideoParams(
+            VideoStreamingCapability hmiCapability,
+            int frameRate, int frameInterval, int width, int height, int bitrate,
+            VideoStreamingFormat[] availableFormats) {
+        ArrayList<VideoStreamingFormat> formats = new ArrayList<>();
+        if (hmiCapability != null && hmiCapability.getSupportedFormats() != null
+                && availableFormats != null) {
+            // supportedFormat is listed in HMI's preferred order
+            for (VideoStreamingFormat supportedFormat : hmiCapability.getSupportedFormats()) {

```

Comment:
It cant be implied that the HMI is returning a list in a preferred order, or any order at all - each HMI implementation could be different. The proxy should have a bit of logic in it to prefer the RTP format over RAW, and fall back if RTP is not in the list.  (which is why the default is still raw in `VideoStreamingParams`).

Reply:
Since right now we support only 2 protocols we could probably just do something like:
```java
List<VideoStreamingFormat> formatList = hmiCapability.getSupportedFormats();
	if (formatList.contains(VIDEO_STREAMING_FORMAT_H264_RTP)){
		// start up preferred RTP format
	} else {
		// iterate list and start up whats available (or just start RAW)
	}
```

Reply:
Thanks for your comment.

Actually, the proposal has description of how the protocol and codec should be selected. HMI notifies the preferred order, and Proxy will try the most preferable format. Please refer to:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0058-video-streaming-capabilities.md

> Head unit's video capabilities are transferred to SDL proxy based on "System Capabilities Query" RPC proposed by SDL-0055 (https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0055-system_capabilities_query.md). The capabilities include:
> * (snip)
> * list of video formats that the head unit supports, in its preferred order. Here, a video format is a combination of "protocol" and "codec".

> When selecting a video format, SDL proxy should pick a format that is most preferable to the head unit and is also supported by the proxy itself. If HMI should reply a negative response with the format, then the proxy picks the second preferable format and try again. SDL proxy can continue this routine until it gets to the end of video format list. If none of the format is accepted (which is unlikely to happen), then the proxy may emit Start Service Control Frame without any video configuration parameters, to fall back to existing sequence.

Honestly speaking, I could implement the preference in Proxy side since we only have two formats (H.264 RTP and H.264 RAW) now. However, we may have more protocols and codecs in future, and it is possible that some HMIs prefer a codec over another codec, or a protocol over another. I wanted to make the spec and implementation flexible enough to support such cases as well.

So my assumption is, all HMIs return the list of supported format in their preferred order.

Reply:
Ok, I do see in the proposal where it says the hmi/core decides the best order to return. However in testing, we did not see that behavior. To test I was editing `hmicapabilities.json` in core and it wasn't responding with RTP first, if both RTP and RAW were present under `videoStreamingCapability`. Please let me know if that was correct behavior or if hmi/core needs to be modified to do the choosing. Thank you

Reply:
Thanks for the info. Let me look into it.
FYI, I found an issue in Core when reading HMI capabilities from a file. I fixed it here https://github.com/smartdevicelink/sdl_core/pull/1763 but I'm not sure if it is related.

Reply:
I verified that following environment works. RTP is selected first, HMI rejects it, then RAW is selected and accepted.

Core: https://github.com/smartdevicelink/sdl_core/pull/1763/commits/412fef738770e97e0c392e5055ddeafb1a1bb9f0 which includes my PR #1763
Capability file: attached [hmi_capabilities.json.zip](https://github.com/smartdevicelink/sdl_android/files/1316608/hmi_capabilities.json.zip)
HMI: https://github.com/smartdevicelink/sdl_hmi/pull/26/commits/79f24a0ecef40ddba02273ff80eddd332c0ec7fa which includes PR #26
Proxy: latest feature/issue_515_rtp_packetizer branch

Please make sure that your SDL app sends GetSystemCapability RPC request with VIDEO_STREAMING type after first OnHMIStatus notification, and start video streaming after receiving GetSystemCapability response. Otherwise, Proxy will not receive HMI capabilities and will use RAW format.

Reply:
Thank you for your quick reply. We will test it out in the next few days

Reply:
@shoamano83 We were able to confirm that it works as you described. Thank you for your help